### ISSUE TITLE
worker validate fails with exception if incoming mmd does not have a defined namespace in the identifier.

### ISSUE THREAD

**Issue by magnarem** [metno/discovery-metadata-catalog-ingestor#156]
I dicovered this error when testing the catalog-rebuilder through dmci.

Here is the debug-log of the error.

```
INFO:werkzeug:127.0.0.1 - - [19/Apr/2023 11:01:03] "POST /v1/insert HTTP/1.1" 500 -
[2023-04-19 11:01:03,210]              dmci.api.worker:290  INFO     XML file title:Direct Broadcast data processed in satellite swath to L1C.
INFO:dmci.api.worker:XML file title:Direct Broadcast data processed in satellite swath to L1C.
[2023-04-19 11:01:03,210]              dmci.api.worker:259  WARNING  metadata_identifier not formed as namespace:UUID
WARNING:dmci.api.worker:metadata_identifier not formed as namespace:UUID
[2023-04-19 11:01:03,210]              dmci.api.worker:96   DEBUG    Identifier namespace: None
DEBUG:dmci.api.worker:Identifier namespace: None
[2023-04-19 11:01:03,210]              dmci.api.worker:97   DEBUG    Environment customization staging
DEBUG:dmci.api.worker:Environment customization staging
[2023-04-19 11:01:03,210]                 dmci.api.app:1744 ERROR    Exception on /v1/insert [POST]
Traceback (most recent call last):
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/magnarem/git/s-enda/metno/discovery-metadata-catalog-ingestor/dmci/api/app.py", line 67, in post_insert
    msg, code = self._insert_update_method_post("insert", request)
  File "/home/magnarem/git/s-enda/metno/discovery-metadata-catalog-ingestor/dmci/api/app.py", line 132, in _insert_update_method_post
    valid, msg, data_ = worker.validate(data)
  File "/home/magnarem/git/s-enda/metno/discovery-metadata-catalog-ingestor/dmci/api/worker.py", line 100, in validate
    if re.search(ns_re_pattern, self._namespace) is None:
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/re.py", line 200, in search
    return _compile(pattern, flags).search(string)
TypeError: expected string or bytes-like object
ERROR:dmci.api.app:Exception on /v1/insert [POST]
Traceback (most recent call last):
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/magnarem/git/s-enda/metno/discovery-metadata-catalog-ingestor/dmci/api/app.py", line 67, in post_insert
    msg, code = self._insert_update_method_post("insert", request)
  File "/home/magnarem/git/s-enda/metno/discovery-metadata-catalog-ingestor/dmci/api/app.py", line 132, in _insert_update_method_post
    valid, msg, data_ = worker.validate(data)
  File "/home/magnarem/git/s-enda/metno/discovery-metadata-catalog-ingestor/dmci/api/worker.py", line 100, in validate
    if re.search(ns_re_pattern, self._namespace) is None:
  File "/home/magnarem/anaconda3/envs/senda/lib/python3.10/re.py", line 200, in search
    return _compile(pattern, flags).search(string)
TypeError: expected string or bytes-like object
```

So the code that fails is both the following lines, since they do the same.

```
logger.debug(re.search(ns_re_pattern, self._namespace))
if re.search(ns_re_pattern, self._namespace) is None:
```

So what happens is that if there are no namespace in the metadata_identifier in the incoming mmd, then the 
`self._namespace` will get the value `None` and then the regexp will fail. See ` dmci.api.worker:96   DEBUG    Identifier namespace: None`

But I thought that line 51 in the `__init__` of the Worker class, that if it is no namespace it  should return empty string, however this not seems to be the case.

A simple fix will to make sure that `self._namespace`  is an empty string or byte-string if no namespace could be extracted from the mmd. Then the regexp changes will not happen as the code is now.

However, how will we handle incoming MMD's that do not have a naming_authority in the identifier?
Should we reject or let it through without any modifications to the identifiers and related_dataset?
But if we assume that all incoming MMDs always will have a naming_authority namespace in the identifier, this will not be an issue.

**Comment by mortenwh** [comment]
I think we agreed that we should require namespaced IDs. So it passes the validation without namespace? It shouldn't.. And I guess that's where we should handle it?

**Comment by magnarem** [comment]
Ok. 
So then we should add a test in `worker.validate()`  after the following test: 
```
 if not isinstance(data, bytes):
            return False, "Input must be bytes type"
```

So something like:
```
 if self._namespace is None:
            return False, "Missing naming authority"
```

And a question. Should those two tests also return back the data or None?

Will this return statements not fail with unpacking since the call for this `worker.validate()` expects three variables back?
Line `132` in `app.py`
```
valid, msg, data_ = worker.validate(data)
```

I can fix this and also update the docstring for the validate() functions as it does not document that data: bytes will be returned.

**Comment by mortenwh** [comment]
yes, you'll need to return three variables. Not sure if it works with None, since you're doing another test after the validation? Maybe just return the same data?

**Comment by charlienegri** [comment]
line 93 contains the check on the namespace 
```
        if valid:
            # Check information content
            valid, msg = self._check_information_content(data)
```
`_check_information_content` contains `_extract_metadata_id` which returns False if no namespace, 

so if False we indeed need to return this before the block `if self._conf.env_string:` 

it should be enough to have
```
valid, msg = self._check_information_content(data)
if not valid :
    return valid, msg, data
```
I can add this to branch 152

**Comment by johtoblan** [comment]
Fixed in #155
